bin/translate-data.ts

```diff
@@ -0,0 +1,117 @@
+#!/usr/bin/env ts-node-script
+
+/** Translate all of the files in `./data/` into properly tagged files in `./dist/data/` */
+
+import path from "path";
+import { promises as fs } from "fs";
+
+import toml from "@iarna/toml";
+
+async function main() {
+  const allData = await DataGroup.gatherFromDir("./data");
+  await allData.writeToDir("./dist/data");
+}
+
+class DataGroup {
+  typeAnnotation: string | null;
+  entries: Array<DataEntry>;
+  children: Record<string, DataGroup>;
+
+  constructor() {
+    this.typeAnnotation = null;
+    this.entries = [];
+    this.children = {};
+  }
+
+  static async gatherFromDir(sourcePath: string): Promise<DataGroup> {
+    const gathered = new DataGroup();
+
+    for (const entry of await fs.readdir(sourcePath, { encoding: "utf-8", withFileTypes: true })) {
+      const fullPath = path.join(sourcePath, entry.name);
+      const { base: nameWithExtension } = path.parse(fullPath);
+
+      if (entry.isDirectory()) {
+        gathered.children[nameWithExtension] = await DataGroup.gatherFromDir(fullPath);
+      } else if (entry.isFile()) {
+        if (nameWithExtension == "_type") {
+          gathered.typeAnnotation = (await fs.readFile(fullPath, { encoding: "utf-8" })).trim();
+          continue;
+        }
+
+        gathered.entries.push(await DataEntry.fromFile(fullPath));
+      }
+    }
+
+    return gathered;
+  }
+
+  async writeToDir(destPath: string): Promise<void> {
+    await fs.mkdir(destPath, { recursive: true });
+
+    for (const entry of this.entries) {
+      await entry.writeIntoDir(destPath);
+    }
+
+    for (const [name, group] of Object.entries(this.children)) {
+      const subdir = path.join(destPath, name);
+      await group.writeToDir(subdir);
+    }
+  }
+}
+
+class DataEntry {
+  name: string;
+  contents: unknown;
+
+  constructor(name: string, contents: unknown) {
+    this.name = name;
+    this.contents = contents;
+  }
+
+  static async fromFile(fullPath: string): Promise<DataEntry> {
+    const { name: nameWithoutExtension, ext } = path.parse(fullPath);
+
+    const fileContents = await fs.readFile(fullPath, { encoding: "utf-8" });
+
+    let parsedContents: unknown;
+    switch (ext) {
+      case ".json": {
+        try {
+          parsedContents = JSON.parse(fileContents);
+        } catch (err) {
+          console.error(`Error: Could not read JSON from ${fullPath}:`);
+          throw err;
+        }
+        break;
+      }
+
+      case ".toml": {
+        try {
+          parsedContents = toml.parse(fileContents);
+        } catch (err) {
+          console.error(`Error: Could not read TOML from ${fullPath}:`);
```

Comment:
It might be useful to throw these errors as is (or include them) since they have line/column information

Reply:
I wish we had Python's re-raise feature that raises an exception without modifying it. Oh well.

The reason I did this structure is because the error message raised by the TOML library didn't know what file it came from, and so it couldn't give useful feedback about what was actually wrong. All it could say is that parsing one of the TOML files broke, and it was up to you to figure out which one it is.